packages/gestalt/src/Typeahead.js

```diff
@@ -0,0 +1,301 @@
+      | SyntheticEvent<HTMLInputElement>,
+  }) => void,
+  onChange?: ({
+    event: SyntheticInputEvent<HTMLInputElement>,
+    value: string,
+  }) => void,
+  onFocus?: ({
+    event: SyntheticFocusEvent<HTMLInputElement>,
+    value: string,
+  }) => void,
+  onSelect?: (?OptionObject) => void,
+  placeholder?: string,
+  searchField?: string,
+  size?: 'md' | 'lg',
+  forwardedRef?: React.Ref<'input'>,
+|};
+
+const Typeahead = (props: Props): Node => {
+  const {
+    data,
+    defaultItem = null,
+    id,
+    label = '',
+    noResultText,
+    onBlur,
+    onChange,
+    onFocus,
+    onSelect,
+    placeholder,
+    searchField = 'label',
+    size,
+    forwardedRef,
+  } = props;
+
+  // Store original data
+  const dataRef = useRef(data);
+
+  // Utility function for filtering data by value
+  const filterOriginalData = filterValue =>
+    dataRef.current.filter(item =>
+      item[searchField].toLowerCase().includes(filterValue.toLowerCase())
+    );
+
+  // Track input value
+  const [search, setSearch] = useState<string>(defaultItem?.label || '');
+
+  // Track the selected item - could be used to see if someone is selecting the same thing again
+  const [selected, setSelected] = useState<OptionObject | null>(defaultItem);
+
+  const [options, setOptions] = useState<OptionObject[]>(
+    filterOriginalData(search)
+  );
+
+  // Ref to the input
+  const inputRef = useRef();
+
+  const [containerOpen, setContainerOpen] = useState<boolean>(false);
+
+  const handleFocus = ({ event, value }) => {
+    // Run focus callback
+    if (onFocus) onFocus({ event, value });
+  };
+
+  const handleBlur = ({ event }) => {
+    // Clear input and reset options when no results
+    if (options.length === 0) {
+      setSearch('');
+      setOptions(dataRef.current);
+    }
+
+    setContainerOpen(false);
+
+    // Run blur callback
+    if (onBlur) onBlur({ event });
+  };
+
+  // Handler for when text is typed
+  const handleChange = ({ event, value }) => {
+    if (containerOpen === false) setContainerOpen(true);
+
+    // Filter the available options using original data
+    const updatedOptions = filterOriginalData(value);
+
+    // Update the avalble options
+    setOptions(updatedOptions);
+
+    // Update the search value
+    setSearch(value);
+
+    // Run onChange callback
+    if (onChange) onChange({ event, value });
+  };
+
+  const handleClear = () => {
+    setSelected(null);
+    setSearch('');
+    if (onSelect) onSelect(null);
+    setContainerOpen(false);
+    if (inputRef.current) inputRef.current.focus();
+  };
+
+  // Handler for when an item is clicked
+  const handleSelect = item => {
+    setSelected(item);
+
+    setSearch(item[searchField]);
+    setContainerOpen(false);
+    if (inputRef.current) inputRef.current.focus();
+    if (onSelect) onSelect(item);
+  };
+
+  const handleScrolling = item => {
+    const selectedNode = [
+      ...document.querySelectorAll('div.typeahead-option'),
```

Comment:
@simkessy this leaks abstraction outside of this component. Instead, could we set a `ref` on the container of options?